server/src/main/java/io/crate/role/TransportPrivilegesAction.java

```diff
@@ -161,5 +171,33 @@ static ApplyPrivsResult applyPrivileges(Metadata.Builder mdBuilder, PrivilegesRe
         return new ApplyPrivsResult(affectedRows, unknownRoleNames);
     }
 
+    private static void validateIsNotUser(Roles roles, RolePrivilegeToApply rolePrivilegeToApply) {
+        for (String roleNameToApply : rolePrivilegeToApply.roleNames()) {
+            if (roles.findRole(roleNameToApply).isUser()) {
+                throw new IllegalArgumentException("Cannot " + rolePrivilegeToApply.state().name() + " a USER to a ROLE");
+            }
+        }
+    }
+
+    @VisibleForTesting
+    static void detectCyclesInRolesHierarchy(Roles roles, PrivilegesRequest request) {
+        if (request.rolePrivilege().state() == PrivilegeState.GRANT) {
```

Comment:
Should we also protect on REVOKE and DENY?

Reply:
DENY role to role is not permitted from analyzer.
I'll think again if REVOKing a role can cause a cycle to appear.

Reply:
I.ve checked also postgres code and I see that the only place this error comes up is here:
https://github.com/postgres/postgres/blob/141752bbb0308a40bdbef3ab0a2ba31162813412/src/backend/commands/user.c#L1753
in https://github.com/postgres/postgres/blob/141752bbb0308a40bdbef3ab0a2ba31162813412/src/backend/commands/user.c#L1688
DelRoleMems: https://github.com/postgres/postgres/blob/141752bbb0308a40bdbef3ab0a2ba31162813412/src/backend/commands/user.c#L1984 doesn't seem to check something.